File: 2023/domaci/15-vyhodnocovac/volfmatous/main.py
import math


def apply_operation(a: float, operation: str, b: float) -> float:
    result = 0

    if operation == "+":
        result = a + b
    elif operation == "-":
        result = a - b
    elif operation == "*":
        result = a * b
    elif operation == "/":
        result = a / b
    elif operation == "^":
        result = math.pow(a, b)

    return result


if __name__ == "__main__":
    n = int(input())

    result = 0

    operation = "+"

    temp1 = 0
    temp1_operation = "+"
    temp1_used = False

    temp2 = 0
    temp2_operation = "+"
    temp2_used = False

    for i in range(n):
        operand = float(input())
        next_operation = "+"

        if i < n - 1:
            next_operation = input()

        if temp2_used:
            temp2 = apply_operation(temp2, temp2_operation, operand)
            temp2_operation = next_operation

        if next_operation == "+" or next_operation == "-" or i >= n - 1:
            if temp1_used:
                if temp2_used:
                    temp1 = apply_operation(temp1, temp1_operation, temp2)
                else:
                    temp1 = apply_operation(temp1, temp1_operation, operand)

                result = apply_operation(result, operation, temp1)
                temp1_used = False
                temp2_used = False
            else:
                result = apply_operation(result, operation, operand)

            operation = next_operation
        elif next_operation == "*" or next_operation == "/":
            if temp1_used:
                if temp2_used:
                    temp1 = apply_operation(temp1, temp1_operation, temp2)
                else:
                    temp1 = apply_operation(temp1, temp1_operation, operand)

                temp1_used = True
                temp1_operation = next_operation
            else:
                temp1_used = True
                temp1 = operand
                temp1_operation = next_operation
        elif next_operation == "^":
            if temp1_used:
                if temp2_used:
                    continue

                if temp1_operation == "^":
                    temp1 = apply_operation(temp1, temp1_operation, operand)
                    temp1_operation = next_operation
                else:
                    temp2_used = True
                    temp2 = operand
                    temp2_operation = next_operation
            else:
                temp1_used = True
                temp1 = operand
                temp1_operation = next_operation

    print(f"{result:g}")

File: 2023/domaci/15-vyhodnocovac/volfmatous/main.js
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});

let n;

let result = 0;

let operation = "+";

let temp1 = 0;
let temp1Operation = "+";
let temp1Used = false;

let temp2 = 0;
let temp2Operation = "+";
let temp2Used = false;

let operand;
let nextOperation;

let loadedOperands = 0;

rl.on("line", (line) => {
    if (n === undefined) {
        n = parseInt(line);
        return;
    }

    if (operand === undefined) {
        loadedOperands++;

        operand = parseFloat(line);

        return;
    }

    nextOperation = line;

    if (temp2Used) {
        temp2 = applyOperation(temp2, temp2Operation, operand);
        temp2Operation = nextOperation;
    }

    if (nextOperation === "+" || nextOperation === "-" || loadedOperands >= n) {
        if (temp1Used) {
            temp1 = applyOperation(temp1, temp1Operation, temp2Used ? temp2 : operand);
            result = applyOperation(result, operation, temp1);

            temp1Used = false;
            temp2Used = false;
        } else {
            result = applyOperation(result, operation, operand);
        }

        operation = nextOperation;
    } else if (nextOperation === "*" || nextOperation === "/") {
        if (temp1Used) {
            temp1 = applyOperation(temp1, temp1Operation, temp2Used ? temp2 : operand);
            temp2Used = false;
        } else {
            temp1Used = true;
            temp1 = operand;
        }

        temp1Operation = nextOperation;
    } else if (nextOperation === "^") {
        if (temp1Used) {
            if (!temp2Used) {
                if (temp1Operation === "^") {
                    temp1 = applyOperation(temp1, temp1Operation, operand);
                    temp1Operation = nextOperation;
                } else {
                    temp2Used = true;
                    temp2 = operand;
                    temp2Operation = nextOperation;
                }
            }
        } else {
            temp1Used = true;
            temp1 = operand;
            temp1Operation = nextOperation;
        }
    }

    operand = undefined;
    nextOperation = undefined;

    if (loadedOperands >= n) {
        rl.close();
    }
});

rl.on("close", () => {
    console.log(result);
});

function applyOperation(a, operation, b) {
    let result = 0;

    switch (operation) {
        case "+":
            result = a + b;
            break;

        case "-":
            result = a - b;
            break;

        case "*":
            result = a * b;
            break;

        case "/":
            result = a / b;
            break;

        case "^":
            result = Math.pow(a, b);
            break;
    }

    return result;
}
